rewrite compiler team documentation (#804)
rust-lang/rust-forge: src/compiler/prioritization.md

# Prioritization
It is important that the compiler team can quickly identify priority issues, hence the establishment
of a prioritization process, described below.

## General Process
1. Ascertain the current status of the issue
1. Try progressing the issue if possible (e.g. request updates from the issue author/reviewer)
1. Is there an MCVE for the issue already?

> 1. Check if it's a regression and label it accordingly (`regression-*` labels)

Author: Fixed!

1. Figure out the area the issue belongs and label it accordingly (`A-*` labels)
1. [Ping notify groups](https://rustc-dev-guide.rust-lang.org/notification-groups/about.html) or
   relevant teams
1. Assign if possible

# Priority Levels
As the compiler team's resources are limited, the primary goal of prioritization is to identify the

> most relevant issues to work on, so that the compiler team can focus on what matters the most.

Author: Fixed!

1. Is there an MCVE for the issue already?
1. Check if it's a regression and label it accordingly (`regression-*` labels)
1. Figure out the area the issue belongs and label it accordingly (`A-*` labels)
1. [Ping notify groups](https://rustc-dev-guide.rust-lang.org/notification-groups/about.html) or
   relevant teams
1. Assign if possible

# Priority Levels
As the compiler team's resources are limited, the primary goal of prioritization is to identify the
most relevant issues to work on, so that the compiler team can focus on what matters the most.

Issues relevant to prioritization are bugs and feature requests that are nominated for
prioritization, by adding the `I-prioritize` label as described below.

## Labels
Labeling an issue as `I-prioritize` starts the prioritization process, which will end by
removing the `I-prioritize` label and appending one of the 4 labels we will discuss below:

- `P-critical`
- `P-high`
- `P-medium`
- `P-low`

Each of these labels defines a strategy the team will adopt regarding:

- The amount of focus a given issue will receive
- How members of the community can get involved

## P-critical

> A `P-critical` issue is a potentially blocker issue. These issues will be raised at the compiler
> team's triage meeting on a weekly basis.

Reviewer: I'm not sure about my wording, I just wanted to convey what kind of "blocker" `P-critical` are. Makes sense?

```suggestion
A `P-critical` is an issue potentially blocking a compiler release (i.e. highly recommended to be solved before a new compiler release). These issues will be raised at the compiler
team's triage meeting on a weekly basis.
```

Reviewer 2: Yeah, I think that's more accurate to how `P-critical` is used (not just compiler release, but the general toolchain release too).

Author: Fixed

soon as possible, and the rest of the team should do their best to help them out if/when applicable.

[critical_downgrade_example]: https://rust-lang.zulipchat.com/#narrow/stream/227806-t-compiler.2Fwg-prioritization/topic/pre-meeting.20triage.202020-04-09.20.2354818

## P-high
`P-high` issues are issues that need attention from the compiler team, but not to the point that
they need to be discussed at every meeting. They can be `P-critical` issues that have a mitigating
condition as defined above, or important issues that aren't deemed blockers.

Because there are too many `P-high` issues to fit in every compiler meeting, they should rather be
handled asynchronously by the team's prioritization, in order to help them move forward. They can
still occasionally be brought up at meetings when it is deemed necessary.

The effectiveness of the team's prioritization will be a direct consequence of the ability to draw
the line between `P-critical` and `P-high` issues. There shouldn't be too many `P-critical` issues
that compiler meetings become unmanageable, but critical issues shouldn't get lost in the list of
`P-high` issues.

> `P-high` issues are issues the teams will mostly work on. We want to make sure they're assigned,
> and keep an eye on them.

Reviewer: ```suggestion
`P-high` issues are issues the teams will mostly work on. We want to make sure they're assigned,
and keep an eye on them. They are routinely reviewed in batches by the compiler team, deciding a possible priority downgrade.
```

Reviewer 2: Could even link an example, e.g. <https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bsteering.5D.202024-11-15.20P-high.20review>.

Author: Fixed

[beta_regressions]: https://github.com/rust-lang/rust/issues?q=is%3Aopen+label%3Aregression-from-stable-to-beta+-label%3AP-critical+-label%3AP-high+-label%3AP-medium+-label%3AP-low+-label%3AT-infra+-label%3AT-libs+-label%3AT-libs-api+-label%3AT-release+-label%3AT-rustdoc
[nightly_regressions]: https://github.com/rust-lang/rust/issues?q=is%3Aopen+label%3Aregression-from-stable-to-nightly+-label%3AP-critical+-label%3AP-high+-label%3AP-medium+-label%3AP-low+-label%3AT-infra+-label%3AT-libs+-label%3AT-libs-api+-label%3AT-release+-label%3AT-rustdoc

The final step prior to generating the agenda is to accept any MCPs. Any MCPs that have had [the `final-comment-period` label][mcp_fcp] 
for more than ten days can be accepted. Remove the `final-comment-period` label and add the `major-change-accepted` label and then
close the issue.

[mcp_fcp]: https://github.com/rust-lang/compiler-team/issues?q=is%3Aissue+is%3Aopen+label%3Amajor-change+label%3Afinal-comment-period

Finally, the meeting agenda can be generated. Clone and build [`triagebot`][triagebot] and set two
environment variables:

- `GITHUB_TOKEN` with a GitHub API token

> - `GOOGLE_API_KEY` with access to the compiler team calendar

Reviewer: We don't use the Google calendar anymore, I take the chance of this refactor to remove this from the docs :slightly_smiling_face: 

```suggestion
```

Author: Fixed
